daqclients-8.2/clients/slowControls/caenv812/caenv812gui.3
'\"
'\" Copyright (c) 2005 Michigan State University All Rights Reserved
'\"  Licensed under the GPL.  See http://www.gnu.org/licenses/gpl.txt
'\"  terms and conditions.
'\"
'\" The definitions below are for supplemental macros used in Tcl/Tk
'\" manual entries.
'\"
'\" .AP type name in/out ?indent?
'\"	Start paragraph describing an argument to a library procedure.
'\"	type is type of argument (int, etc.), in/out is either "in", "out",
'\"	or "in/out" to describe whether procedure reads or modifies arg,
'\"	and indent is equivalent to second arg of .IP (shouldn't ever be
'\"	needed;  use .AS below instead)
'\"
'\" .AS ?type? ?name?
'\"	Give maximum sizes of arguments for setting tab stops.  Type and
'\"	name are examples of largest possible arguments that will be passed
'\"	to .AP later.  If args are omitted, default tab stops are used.
'\"
'\" .BS
'\"	Start box enclosure.  From here until next .BE, everything will be
'\"	enclosed in one large box.
'\"
'\" .BE
'\"	End of box enclosure.
'\"
'\" .CS
'\"	Begin code excerpt.
'\"
'\" .CE
'\"	End code excerpt.
'\"
'\" .VS ?version? ?br?
'\"	Begin vertical sidebar, for use in marking newly-changed parts
'\"	of man pages.  The first argument is ignored and used for recording
'\"	the version when the .VS was added, so that the sidebars can be
'\"	found and removed when they reach a certain age.  If another argument
'\"	is present, then a line break is forced before starting the sidebar.
'\"
'\" .VE
'\"	End of vertical sidebar.
'\"
'\" .DS
'\"	Begin an indented unfilled display.
'\"
'\" .DE
'\"	End of indented unfilled display.
'\"
'\" .SO
'\"	Start of list of standard options for a Tk widget.  The
'\"	options follow on successive lines, in four columns separated
'\"	by tabs.
'\"
'\" .SE
'\"	End of list of standard options for a Tk widget.
'\"
'\" .OP cmdName dbName dbClass
'\"	Start of description of a specific option.  cmdName gives the
'\"	option's name as specified in the class command, dbName gives
'\"	the option's name in the option database, and dbClass gives
'\"	the option's class in the option database.
'\"
'\" .UL arg1 arg2
'\"	Print arg1 underlined, then print arg2 normally.
'\"
'\" RCS: @(#) $Id$
'\"
'\"	# Set up traps and other miscellaneous stuff for Tcl/Tk man pages.
.if t .wh -1.3i ^B
.nr ^l \n(.l
.ad b
'\"	# Start an argument description
.de AP
.ie !"\\$4"" .TP \\$4
.el \{\
.   ie !"\\$2"" .TP \\n()Cu
.   el          .TP 15
.\}
.ta \\n()Au \\n()Bu
.ie !"\\$3"" \{\
\&\\$1	\\fI\\$2\\fP	(\\$3)
.\".b
.\}
.el \{\
.br
.ie !"\\$2"" \{\
\&\\$1	\\fI\\$2\\fP
.\}
.el \{\
\&\\fI\\$1\\fP
.\}
.\}
..
'\"	# define tabbing values for .AP
.de AS
.nr )A 10n
.if !"\\$1"" .nr )A \\w'\\$1'u+3n
.nr )B \\n()Au+15n
.\"
.if !"\\$2"" .nr )B \\w'\\$2'u+\\n()Au+3n
.nr )C \\n()Bu+\\w'(in/out)'u+2n
..
.AS Tcl_Interp Tcl_CreateInterp in/out
'\"	# BS - start boxed text
'\"	# ^y = starting y location
'\"	# ^b = 1
.de BS
.br
.mk ^y
.nr ^b 1u
.if n .nf
.if n .ti 0
.if n \l'\\n(.lu\(ul'
.if n .fi
..
'\"	# BE - end boxed text (draw box now)
.de BE
.nf
.ti 0
.mk ^t
.ie n \l'\\n(^lu\(ul'
.el \{\
.\"	Draw four-sided box normally, but don't draw top of
.\"	box if the box started on an earlier page.
.ie !\\n(^b-1 \{\
\h'-1.5n'\L'|\\n(^yu-1v'\l'\\n(^lu+3n\(ul'\L'\\n(^tu+1v-\\n(^yu'\l'|0u-1.5n\(ul'
.\}
.el \}\
\h'-1.5n'\L'|\\n(^yu-1v'\h'\\n(^lu+3n'\L'\\n(^tu+1v-\\n(^yu'\l'|0u-1.5n\(ul'
.\}
.\}
.fi
.br
.nr ^b 0
..
'\"	# VS - start vertical sidebar
'\"	# ^Y = starting y location
'\"	# ^v = 1 (for troff;  for nroff this doesn't matter)
.de VS
.if !"\\$2"" .br
.mk ^Y
.ie n 'mc \s12\(br\s0
.el .nr ^v 1u
..
'\"	# VE - end of vertical sidebar
.de VE
.ie n 'mc
.el \{\
.ev 2
.nf
.ti 0
.mk ^t
\h'|\\n(^lu+3n'\L'|\\n(^Yu-1v\(bv'\v'\\n(^tu+1v-\\n(^Yu'\h'-|\\n(^lu+3n'
.sp -1
.fi
.ev
.\}
.nr ^v 0
..
'\"	# Special macro to handle page bottom:  finish off current
'\"	# box/sidebar if in box/sidebar mode, then invoked standard
'\"	# page bottom macro.
.de ^B
.ev 2
'ti 0
'nf
.mk ^t
.if \\n(^b \{\
.\"	Draw three-sided box if this is the box's first page,
.\"	draw two sides but no top otherwise.
.ie !\\n(^b-1 \h'-1.5n'\L'|\\n(^yu-1v'\l'\\n(^lu+3n\(ul'\L'\\n(^tu+1v-\\n(^yu'\h'|0u'\c
.el \h'-1.5n'\L'|\\n(^yu-1v'\h'\\n(^lu+3n'\L'\\n(^tu+1v-\\n(^yu'\h'|0u'\c
.\}
.if \\n(^v \{\
.nr ^x \\n(^tu+1v-\\n(^Yu
\kx\h'-\\nxu'\h'|\\n(^lu+3n'\ky\L'-\\n(^xu'\v'\\n(^xu'\h'|0u'\c
.\}
.bp
'fi
.ev
.if \\n(^b \{\
.mk ^y
.nr ^b 2
.\}
.if \\n(^v \{\
.mk ^Y
.\}
..
'\"	# DS - begin display
.de DS
.RS
.nf
.sp
..
'\"	# DE - end display
.de DE
.fi
.RE
.sp
..
'\"	# SO - start of list of standard options
.de SO
.SH "STANDARD OPTIONS"
.LP
.nf
.ta 5.5c 11c
.ft B
..
'\"	# SE - end of list of standard options
.de SE
.fi
.ft R
.LP
See the \\fBoptions\\fR manual entry for details on the standard options.
..
'\"	# OP - start of full description for a single option
.de OP
.LP
.nf
.ta 4c
Command-Line Name:	\\fB\\$1\\fR
Database Name:	\\fB\\$2\\fR
Database Class:	\\fB\\$3\\fR
.fi
.IP
..
'\"	# CS - begin code excerpt
.de CS
.RS
.nf
.ta .25i .5i .75i 1i
..
'\"	# CE - end code excerpt
.de CE
.fi
.RE
..
.de UL
\\$1\l'|0\(ul'\\$2
..
.TH caenv812gui 3 "" Tcl "NSCL DAQ Tcl support"
.BS
'\" Note:  do not modify the .SH NAME line immediately below!
.SH NAME
caenv812gui \- Megawidget control panel for the CAEN V812 CFD.
.SH SYNOPSIS
.DS
\fBpackage require caenv812
caenv812\fI pathName ?options?\fR
.DE

.BE

.SH DESCRIPTION
.PP
caenv812gui supplies a megawidget that is a control panel for the
CAEN V812 constant fraction discriminator.  The megawidget is the
same as that used by the caenv812control application, and is available
for use in custom user control applications.
.SH "WIDGET-SPECIFIC OPTIONS"
.OP -crate "" ""
Specifies the VME crate in which the module controlled by this
widget lives.  If this is not supplied when constructing the widget,
it defaults to 0.  You may not configure the \fB-crate\fR after the
widget has been constructed as the widget is bound to the module when it
is first constructed.
.OP -base "" ""
Required option at construction time.  The argument to this option is the
base address of the module.  There is no default.  The \fB-base\fR option
must be supplied at construction time and cannot be reconfigured after the
widget is initially constructed.
.OP -disable "" ""
Creating or configuring with this switch allows parts of the user interface
to be disabled.  The \fB-disable\fR option accepts as an argument
a list of the interface items to disable.
These are not cumulative across multiple uses of \fB-disable\fR.  Legitimate
list items are \fBwidths\fR, \fBmajority\fR, and \fBdeadtimes\fR  This
configuration option allows the GUI to be used for other compatible
discrimnator modules that don't have all of these controls.  For example, the
CAEN V895 leading edge discriminator by definition does not have a deadtime.
.OP -name "" ""
Required option at construction time.  The argument to this option is the
name of the module.  This module name is associated with the module address
in order to interface the widget with the device support packages CFD812 and CFDState.
This option cannot be reconfigured after construction.
.OP -command "" ""
The parameter to this option is a script to be executed when the control
panel changes a setting in the module.  This allows the user to layer
e.g. logging on top of the control panel widget.   Multiple \fB-command\fR
options are not cumulative.  To remove a previously set \fB-command\fR script,
supply the \fB-command\fR option with the script "".

The user script is invoked with a three element list appended to it of the form:
{action id newvalue}.
.DS
Where:
    action  is text describing what changed and is one of:
            threshold  - A threshold changed.
            width      - A width changed.
            deadtime   - A deadtime changed.
            majority   - The majority threshold changed.
            mask       - The enables mask.

    newvalue is a new control value associated with the change.

    id       Describes which item was changed.  This will be a channel number
             for thresholds, 0 or 1 for widths/deadtimes and an empty string for
             majority threshold and the mask.
.DE
.OP -alarmscript "" ""
The parameter to this option is a script that will be executed when
the control panel detects an alarm condition.  The alarm script will
be passed a single parameter which is a human readable alarm string.
Note that caenv812gui does not currently define any alarms, however this
option is provided for compatibility with devices that do support alarms.
.SH "WIDGET COMMANDS"
.TP
\fIpathName\fB getProperties\fR
Retrieves a property list that describes the current device settings.  Note that given the
write\-only nature of the hardware, if multiple processes exercise control over the same
physical module, this property list will only be the widgets understanding of what the
settings would be if nobody else were meddling with the device.

Property lists are a device independent way to represent module settings.  A property list
is a Tcl list of two element sublists, called properties.  Each property consists of a
property name and value. For the properties supported by the caenv812gui widget,
see \fBPROPERTIES\fR below.
.TP
\fIpathName\fB setProperties\fI propertyList\fR
Requests that the gui process the \fIpropertyList\fR parameter and set the module accordingly.
.SH "PROPERTIES"
The following properties are supported by the caenv812gui:
.TP
\fBthreshold0 .. threshold15\fR
The thresholds of the module's 16 channels.
.TP
\fBwidth0, width1\fR
The widths of the lower and upper bank of module channels.  Bank 0 controls
channels 0\-7 while bank 1 controls channels 8\-15.

CHANGELOG
NSCLDAQ-7.1

7.1-005 3/31/03
    -  Allow Readout/Active.cpp Active::EndRun to be harmlessly called
       when the run is not in progress.

7.1-002-004 -
	Fixed misc. problems with Stager and with ReadoutGui
        to make these more production quality.

7.1-001  2/11/03:
  - ./configure.in  - Fix an error in the search for the SBS/Bit3 headers.
    the script would fail if the initial set of directories did not contain
    the headers...even if the secondary search set did.
  - VMEApi/Makefile.am
    VMEApi/SBSBIT3API.h  - this file was mis-named and should have been called
	SBSBit3API.h  (note difference in capitalization).  The file has been
        renamed and Makefile.am fixed up to produce an appropriate tarball.
  - Stager/directories.tcl - Modified to support a BUFFERSIZE environment
	variable that can be used to modify the behavior of
	ExpFileSystem::GenRunFile in the event the user has overidden the
	default daq buffersize

7.1-002 3/2/03:
-   Correct the name of the Vme API library.
-   No longer set the busy when scalers are being read as that can
    cause false triggers in the case where the user is using the busy
    as the trigger signal.

7.1-003 3/17/03
	ReadoutGUI defect fix:  If a link to the event file already exists
	in the ~/experiment/current directory at the start of a recorded run,
	the run start fails.  Fixed by removing such links if required.

	ReadoutGUI defect fix:  If readout is being recompiled as it runs,
	ReadoutGUI can hang or readout can die.  Fixed by copying readout
	program to 'private hidden' directory first.

	Added production readout software to the autoconf build, along with
	its documentation.

7.1-004 3/23/03
	This release adds the SBS/Bit3 device driver to the build.
	 SBS/Bit3 has kindly released the source code of this driver for
	open distribution so long as redistribution includes acknowledgement
	of the SBS Technologies copyright.

	From now on, it will be necessary to be root to build the system
	in order to get the device driver properly installed.

	Changes:
	- Minor bug fixes.
	- SBS/Bit3 device driver included in the build.

7.1-005 3/31/03
    -  Allow Readout/Active.cpp Active::EndRun to be harmlessly called
       when the run is not in progress.

NSCLDAQ-7.2

7.2    ??/??/??
    - Support multi-crate configurations of the CAENCard and CAENChain
      modules.
    - Support for the CAEN V890 multihit tdc.
    - Many stager bugs fixed.  to alloow multiple hits

7.2.2 12/05/03
   - Abstracting the device support libraries from the Readout software.

7.2.2-001 12/05/03
   - Stager: At install time create a package InstallRoot that allows
     us to figure out where we're installed (needed to reliably locate
     eventlog e.g.).

NSCLDAQ- 7.3

7.3    12/05/03
    - Incorporate the CAEN V812 CFD GUI developed by Ron Fox,
      Mike Famiano, and Anton Zalessov:
       $(prefix)/contrib/caenv812.
    - Incorporate scripted readout engine developed initially for
      LLNL, LBL then SEE and RPI by Ron Fox.
    - Added bunches of stuff to the contrib directory for scriptable readout.


7.3-005 3/3/04
    - Did a merge with the 7.4 branch for:
       framework
       controlpush ...
       ProductionReadout
       configure.in
       Makefile.am
     in order to get the controlpush functionality and readout fixes that
     go along with it.
    - Fix incorrect base address for vme trigger/status module was
      0xf000 sb. 0x444400
7.3-006 3/10/04
   - Merge production readout with 7.4 branch to incorporate fix for
     issue 116: Title, run variable and state variable values that are too
     large cause all sorts of grief.
7.3-007 3/24/04
   - Merge Scaler client from 7.4 to catch a long standing issue I'd
     forgotton about: Update needs to be called even if there are no
     scalers (if only to keep track of run time).
   - Merge Production readout with 7.4 branch to incorporate fix for
     issue 106: Elapsed run time incorrect.
7.3-008 3/31/04
   - Merge in the fixes to the Stager/ReadoutGui that constitute the fix
     for issue 108: Event files can be ovewritten with no prompt.
   - Merge in fix for production readout - stabilization of scaler readout timings.
7.3-008 4/12/04
   - Added build stamp to the installation procedure
   - Fixed an incompatibility with tcl 8.4 and cratelocator.tcl
   - Fixed controlpush - plc channels need extra time to connect
   - Added # escaping to controlpush and to production readout
7.3-009 4/18/04
   - Modify Stager so that it can record tape on a different node than
     it records ftp data.
   - Issue 112: Warn the user if the scaler client is started with the default
     data source.  At the NSCL this is an unusual occurence.  Note that
     the warning will be sent via an Xwindows window if possible or if not
     via stderr.
7.3-010

clients/slowControls/caenv812/caenv812Gui.tcl
#
#    This software is Copyright by the Board of Trustees of Michigan
#    State University (c) Copyright 2005.
#
#    You may use this software under the terms of the GNU public license
#    (GPL).  The terms of this license are described at:
#
#     http://www.gnu.org/licenses/gpl.txt
#
#     Author:
#             Ron Fox
#            NSCL
#            Michigan State University
#            East Lansing, MI 48824-1321
#


package require Tk
package require snit

#  Support packages for the CFD:

package require CFD812   2.0
package require CFDState 2.0


package provide caenv812Gui 1.0

# This is a GUI for the CAEN V812 constant fraction discriminator.
# unlike the previous >BIG FOOTPRINT< version, this is a small footprint
# version that allows you to control a single channel at a time.
# The GUI layout is as follows:
#
#   +--------------------------------------------------------------------+
#   |    CAENV812 @          base_address  Serial Number: serial_number  |
#   |                     +------------------------------+               |
#   |                     |  Channel:  |  |^             |               |
#   |                     |            |  |V             |               |
#   |                     |                              |               |
#   |                     |  Threshold |  |^             |               |
#   |                     |            |  |V             |               |
#   |                     |                              |               |
#   |                     | Enable [ ]  <Set All>        |               |
#   |                     +------------------------------+               |
#   | +------------------------+             +-----------------------+   |
#   | |Widths                  |             | Deadtimes             |   |
#   | |Channels 0-7  |  |^     |             | Channels 0-7  | |^    |   |
#   | |Channels 8-15 |  |^     |             | Channels 8-15 | |V    |   |
#   | +------------------------+             +-----------------------+   |
#   |                      Majority Threshold | |^                       |
#   +--------------------------------------------------------------------+
#
#  Options:
#      -crate     The VME crate that has the module.
#      -base      The base address of the module in the crate.
#      -name      The name of the module.  Used to set the name of the state
#                 of the module in CFDState.  If not supplied, we use the object
#                 name at the time of construction.
#      -command   Supplies a script that will be invoked to describe changes done
#                 the script will have a list appended to it of the following form:
#                 {action id newvalue}
#                 Where:
#                     action   is text describing what changed:
#                              threshold  - A threshold changed.
#                              width      - A width changed.
#                              deadtime   - A deadtime changed.
#                              majority   - The majority threshold changed.
#                              mask       - The enables mask.
#
#                     newvalue is a new control value associated with the
#                              change.
#                     id       Describes which item was changed.  This will be a channel number
#                              for thresholds, 0 or 1 for widths/deadtimes and an empty string for
#                              majority threshold and the mask.
#      -alarmscript  Supplies a script that will be invoked if the module has any alarms to report.
#                    Note that this module has no alarm conditions however this option is provided
#                    to unify the interface of all control widgets
#      -disable thing Allows elements of the display to be disabled (ghosted).  thing can be a list of
#                      widths   - Disable width settings
#                      majority - Disable majority settings.
#                      deadtimes- Disable dead time settings.
#                     typically only deadtimes is used as that's needed for the V895 leading edge
#                     discriminator. 
#
# Properties:
#     threshold0 .. threshold15  - The thresholds of the module's 16 channels.
#     width0, width1             - The widths of the lower and upper 1/2 of the channels.
#     deadtime0, deadtime1       - The deadtimes of the lower and upper 1/2 of the channels.
#     majority                   - The majority threshold.
#     enables                    - The enables mask.
# Standard methods:
#     getProperties              - Returns the current property list.
#     setProperties              - Sets the module's current property lists.
#
snit::widget cfdv812Gui {
    option -crate 0
    option -base  {}
    option -name  {}
    option -command {}
    option -alarmscript {}
    option -disable {}


    variable cfd;        # holds the handle to the cfd object.

    constructor args {
        set options(-name) $self
        $self configurelist $args


        # Get the device and state 'objects' made.

        if {[catch {$self createDevice} msg]} {
	    error "controller creation failed: $msg"
	}
        $self createState

        # Since all that worked we can now layout the GUI.

        # all the labels at the top of the window:

        label $win.baselabel   -text {Base Addr: }
        label $win.baseaddress -text $options(-base)

        label $win.seriallabel -text {Serial Number: }
        label $win.serialnum   -text [::CFD812::GetSerialNumber $cfd]

        # The channel selector and threshold box:
        #

        set channel [frame $win.channel -relief groove -borderwidth 3]
        label   $channel.chanlabel      -text {Channel }
        spinbox $channel.channel        -from 0 -to 15 -increment 1 -width 2   \
                                        -command [mymethod changeChannel $channel.channel]
        label   $channel.threshlabel    -text {Threshold (V)}
        spinbox $channel.threshold      -from -255 -to -1 -increment 1 -width 4 \
                                        -command [mymethod setThreshold $channel.channel $channel.threshold]
        checkbutton $channel.enable    -text {Enable} -command [mymethod flipDisable $channel.channel]
        button      $channel.setall     -text {Set All} \
                                        -command [mymethod setAll $channel.threshold]

        #   The Widths:

        set widths [frame $win.widths -relief groove -borderwidth 3]
        label $widths.label    -text {Width (Arb)}
        label $widths.w07label -text {Channels 0-7}
        spinbox $widths.w07    -from 0 -to 255 -increment 1 -width 3 \
                                -command [mymethod setWidth $widths.w07 0]
        label $widths.w815label -text {Channels 8-15}
        spinbox $widths.w815    -from 0 -to 255 -increment 1 -width 3 \
                                -command [mymethod setWidth $widths.w815 1]


        #  The deadtimes.

        set deadtimes [frame $win.deadtime -relief groove -borderwidth 3]
        label $deadtimes.label    -text {Deadtime (Arb)}
        label $deadtimes.d07label -text {Channels 0-7}
        spinbox $deadtimes.d07    -from 0 -to 255 -increment 1 -width 3 \
                                -command [mymethod setDeadtime $deadtimes.d07 0]
        label $deadtimes.d815label -text {Channesl 8-15}
        spinbox $deadtimes.d815   -from 0 -to 255 -increment 1 -width 3  \
                                -command [mymethod setDeadtime $deadtimes.d815 1]

        # Majority threshold:

        label   $win.majoritylabel -text {Majority Threshold}
        spinbox $win.majority      -from 1 -to 16 -increment 1 -width 3 \
                                -command [mymethod setMajorityLevel $win.majority]

        # Setup the geometry:

        #  The channel subframe.

        grid $channel.chanlabel   $channel.channel
        grid $channel.threshlabel $channel.threshold
        grid $channel.enable      $channel.setall

        # The widths subframe.

        grid $widths.label              x
        grid $widths.w07label     $widths.w07  -sticky w
        grid $widths.w815label    $widths.w815 -sticky w

        # The deadtimes subframe

        grid $deadtimes.label           x
        grid $deadtimes.d07label  $deadtimes.d07  -sticky w
        grid $deadtimes.d815label $deadtimes.d815 -sticky w

        # The overall layout.


        grid $win.baselabel $win.baseaddress $win.seriallabel $win.serialnum
        grid      x         $win.channel           -              x -sticky ew
        grid $widths             -           $deadtimes           - -sticky w
        grid      x         $win.majoritylabel $win.majority      x

        # Set bindings so that Enter and Focus out will also change values.
        #
        bind $channel.threshold <FocusOut> [mymethod setThreshold $channel.channel $channel.threshold]
        bind $channel.threshold <Return>   [mymethod setThreshold $channel.channel $channel.threshold]

        bind $widths.w07        <FocusOut> [mymethod setWidth $widths.w07 0]
        bind $widths.w07        <Return>   [mymethod setWidth $widths.w07 0]

        bind $widths.w815       <FocusOut> [mymethod setWidth $widths.w815 1]
        bind $widths.w815       <Return>   [mymethod setWidth $widths.w815 1]

        bind $deadtimes.d07     <FocusOut> [mymethod setDeadtime $deadtimes.d07 0]
        bind $deadtimes.d07     <Return>   [mymethod setDeadtime $deadtimes.d07 0]
        bind $deadtimes.d815    <FocusOut> [mymethod setDeadtime $deadtimes.d815 1]
        bind $deadtimes.d815    <Return>   [mymethod setDeadtime $deadtimes.d815 1]

        bind $win.majority      <FocusOut> [mymethod setMajorityLevel $win.majority]
        bind $win.majority      <Return>   [mymethod setMajorityLevel $win.majority]

        $self selectChannel 0

	$self setDisables;		# Disable any elements requested.
    }
    # destructor:
    #    Called when one of these widgets is destroyed.
    #

    destructor {
        catch {$self destroyDevice}
        catch {$self destroyState}

    }
    #
    # Process the -disable switch only if the constructor has completed.. otherwise
    # the constructor will process that switch
    #
    onconfigure -disable value {
	set options(-disable) $value

	if {[winfo exists $win.majority]} {
	    $self setDisables
	}
    }

    # Private method to modify the GUI to match the set of disables programmed
    # by the -disable option.  This can be a list containing any of the
    # following words:
    #    widths     - Disables the width widget set.
    #    majority   - Disables the majority widget set.
    #    deadtimes  - Disables the deadtime widget set.
    #
    method setDisables {} {
	foreach word [list widths majority deadtimes] {
	    if {[lsearch -exact $options(-disable) $word] == -1} {
		set state normal
	    } else {
		set state disabled
	    }
	    $self  set${word}State $state
	}
    }

    # State setting methods setwidthsState, setmajorityState and setdeadtimesState
    # each takes a state setting to apply to the specified widget cluster.
    #
    method setwidthsState setting {
	$win.widths.w07  config  -state $setting
	$win.widths.w815 config  -state $setting
    }

    method setmajorityState setting {
	$win.majority config -state $setting
    }

    method setdeadtimesState setting {
	$win.deadtime.d07  config -state $setting
	$win.deadtime.d815 config -state $setting
    }

    # setThreshold channel threshold
    #        Sets the threshold to a new value.
    # Parameters:
    #    channel    - spinbox widget that holds the current channel
    #    threshold  - spinbox widget that holds the current threshold
    # NOTE:
    #    Actualy the widgets can be anything that supports a get method.
    #
    method setThreshold {channel threshold} {
        set chanNum     [$channel   get]
        set threshValue [$threshold get]
        $self setChannelThreshold $chanNum $threshValue
    }
    # setAll threshold
    #     Sets the threshold for all channels to a new value.
    #
    # Parameters:
    #   threshold  - A widget that supports a get method that will return the
    #                requested threshold.
    #
    method setAll threshold {
        set threshValue  [$threshold get]
        for {set chanNum 0} {$chanNum < 16} {incr chanNum} {
            $self setChannelThreshold $chanNum $threshValue
        }
    }
    # createDevice
    #       Create the initial cfd device 'object.'
    #
    method createDevice {} {
        set cfd [::CFD812::Map $options(-base) $options(-name) $options(-crate)]

    }
    # createState
    #      Create the intial cfd state 'object'
    #
    method createState  {} {
        ::CFDState::Create $options(-name)
    }
    # destroyDevice
    #     Destroy the cfd device object.
    #
    method destroyDevice {} {
        ::CFD812::Unmap $options(-name)
    }
    # destroyState
    #    Destroy the state object.
    #
    method destroyState {} {
        ::CFDState::Destroy $options(-name)
    }
    # setChannelThreshold channel value (private)
    #     Set the threshold of a channel to value
    # Parameters:
    #    channel  - The >number< of the channel.
    #    value    - the >value< of the threshold.
    #
    method setChannelThreshold {channel value} {
        ::CFD812::SetThreshold   $cfd            $channel $value
        ::CFDState::SetThreshold $options(-name) $channel $value
        $self log threshold $channel $value
    }
    # selectChannel channel
    #       Change the channel to the one in the spinbox.  We
    #       fetch the new channel's threshold and enable state from
    #       our state object (the cfd is write-only) and set the
    #       corresponding gui widgets.
    # Parameters:
    #   channel  - The channel we are switching to
    method selectChannel channel {
        set threshValue [::CFDState::GetThreshold $options(-name) $channel]
        set mask        [::CFDState::getMask      $options(-name)]

        $win.channel.threshold set $threshValue
        if {($mask & [expr 1 << $channel]) != 0} {
            $win.channel.enable select

        } else {
            $win.channel.enable deselect
        }

    }
    # changeChannel spinbox
    #         Change the channel to the one on the spinbox.. We fetch the
    #         channel and call selectChannel to do the actual work.
    # Parameters
    #   spinbox  - A widget that on being given a 'get' method will return
    #              the desired channel number.
    #
    method changeChannel spinbox {
        set channel [$spinbox get]
        $self selectChannel $channel
    }
    # setModuleEnables mask
    #        Set the module's enable mask
    # Parameters:
    #   mask  - The new mask.
    #
    method setModuleEnables mask {
        ::CFD812::SetMask   $cfd $mask
        ::CFDState::setMask $options(-name) $mask
        $self log mask {} [format "0x%x" $mask]
    }
    # flipDisable spinbox
    #         Toggles the state of the disable bit from the one memorized in the mask.
    # Parameters:
    #   spinbox  - A widget which, when get is invoked, returns the channel.
    #
    method flipDisable spinbox {
        set channel  [$spinbox get]
        set mask     [::CFDState::getMask $options(-name)]
        set chanbit  [expr 1 << $channel]
        set mask     [expr $mask ^ $chanbit]
        if {($mask & $chanbit) != 0} {
            $win.channel.enable select
        } else {
            $win.channel.enable deselect
        }
        $self setModuleEnables $mask

    }
    # setBankWidth bank value
    #        Set the width of a bank of 8 channels.
    # Parameters:
    # bank    - The bank number (0 for 0-7, 1 for 8-15).
    # width   - The new width value in arbitrary units.
    #
    method setBankWidth {bank width} {
        ::CFD812::SetWidth $cfd $bank $width
        ::CFDState::SetWidth $options(-name) $bank $width
        $self log width $bank $width
    }
    # setWidth spinbox bank
    #        Updates the width of a discriminator bank.  See Parameters
    #        below for more on what constitutes a bank.
    # Parameters:
    #  spinbox   - A widget which, when the 'get' method is applied to it returns
    #              the new width.
    #  bank      - 0 if the width is for channels 0-7, 1 if for channels 8-15. Groups of
    #              8 channels share a single width.
    #
    method setWidth {spinbox bank} {
        set width [$spinbox get]
        $self setBankWidth $bank $width
    }
    # setBankDeadtime bank deadtime
    #      Set the deadtime for a bank of the discriminator.
    # Parameters:
    #    bank    - The bank to set (0 for chans 0-7, 1 for chans 8-15).
    #    deadtime- The new deadtime in arbitrary units.
    #
    method setBankDeadtime {bank deadtime} {
        ::CFD812::SetDeadtime    $cfd $bank $deadtime
        ::CFDState::SetDeadtime  $options(-name) $bank $deadtime
        $self log deadtime $bank $deadtime
    }
    # setDeadtime spinbox bank
    #       Updates the deadtime value of a discriminator bank.  See the
    #       Parameters comments for a definition of discriminator bank.
    # Parameters:
    #    spinbox  - A widget which, when the 'get' method is applied to it
    #               returns the new deadtime.
    #    bank     - The discriminator bank to set.  Bank 0 is channels 0-7,
    #               Bank 1 is channels 8-15.
    #
    method setDeadtime {spinbox bank} {
        set deadtime [$spinbox get]
        $self setBankDeadtime $bank $deadtime

    }
    # setModuleMajority majority
    #       Set the threshold for the majority logic unit.
    # Parameters
    #   majority - The new majority threshold.
    #
    method setModuleMajority majority {
        ::CFD812::SetMultiplicityThreshold $cfd $majority
        ::CFDState::SetMultiplicity $options(-name) $majority
        $self log majority {} $majority
    }
    # setMajorityLevel spinbox
    #        Set a new majority level for the module.
    # Parameters:
    #   spinbox - A widget that supports the get method to return its value.
    #
    method setMajorityLevel spinbox {
        set majority [$spinbox get]
        $self setModuleMajority $majority
    }
    # log  what which value
    #     Logs a module change.   If -options(-command) is non-null the
    #     callback script is constructed and invoked.
    # Parameters (see the -command option description for more detail):
    #   what   - What has changed.
    #   which  - Which one of them has changed
    #   value  - New value of the thing that changed.
    #
    method log {what which value} {
        set script $options(-command)
        if {$script ne ""} {
            eval $script [list $what $which $value]
        }
    }
    # setProperties list
    #      Asks the module to set values from the attached property list.
    #      see the module header for more information about the supported properties.
    # Parameters:
    #   list  - a list of {name value} pairs that make up the property list.
    #
    method setProperties {list} {
        foreach item $list {
            set property [lindex $item 0]
            set value    [lindex $item 1]
            switch -glob $property {
                threshold* {
                    scan $property "threshold%d" channel
                    $self setChannelThreshold $channel $value
                }
                width* {
                    scan $property "width%d" bank
                    $self setBankWidth $bank $value
                }
                deadtime* {
                    scan $property "deadtime%d" bank
                    $self setBankDeadtime $bank $value
                }
                majority {
                    $self setModuleMajority $value
                }
                enables {
                    $self setModuleEnables $value
                }
                default {
                    error "Bad property in CAEN V812 propertylist: $property $value"
                }
            }
        }
        $self selectChannel [$win.channel.channel get]
        $win.widths.w07  set [::CFDState::GetWidth $options(-name) 0]
        $win.widths.w815 set [::CFDState::GetWidth $options(-name) 1]

        $win.deadtime.d07  set [::CFDState::GetDeadtime $options(-name) 0]
        $win.deadtime.d815 set [::CFDState::GetDeadtime $options(-name) 1]

        $win.majority       set [::CFDState::GetMultiplicity $options(-name)]
    }
    # getProperties
    #      Returns the module's current property list.
    #
    method getProperties {} {
        for {set i 0} {$i < 16} {incr i} {
            set property [format threshold%d $i]
            set value    [::CFDState::GetThreshold $options(-name) $i]
            lappend properties [list $property $value]
        }
	set mask [::CFDState::getMask $options(-name)]
	lappend properties [list "enables" $mask]
        foreach i {0 1} {
            set property [format width%d $i]
            set value    [::CFDState::GetWidth $options(-name) $i]
            lappend properties [list $property $value]
            set property [format deadtime%d $i]
            set value    [::CFDState::GetDeadtime $options(-name) $i]
            lappend properties [list $property $value]
        }
        lappend properties [list majority [::CFDState::GetMultiplicity $options(-name)]]

        return $properties
    }
}
